cloud-controller-manager/do/doks_lb_validator.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2023 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package do
+
+import (
+	"context"
+	"fmt"
+	"github.com/digitalocean/godo"
+	"github.com/go-logr/logr"
+	"golang.org/x/oauth2"
+	v1 "k8s.io/api/core/v1"
+	"net/http"
+	"os"
+	"strconv"
+
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// LBService represents the v1.service lb object data
+type LBService struct {
+	APIVersion string
+	Kind       string
+	Metadata   Metadata
+	Spec       map[string]interface{}
+	Selector   map[string]interface{}
+	Type       string
+}
+
+// Metadata represents the metadata field in the lb service object
+type Metadata struct {
+	Annotations interface{} `json:"annotation data,omitempty"`
+	Name        string
+	Namespace   string
+}
+
+// DOKSLBServiceValidator validates service type LB
+type DOKSLBServiceValidator struct {
+	decoder *admission.Decoder
+	Log     logr.Logger
+	gClient *godo.Client
+}
+
+func initDOClient() (doClient *godo.Client, err error) {
+	token := os.Getenv(doAccessTokenEnv)
+
+	opts := []godo.ClientOpt{}
+
+	opts = append(opts, godo.SetUserAgent("digitalocean-webhook-server/"+version))
+
+	if token == "" {
+		return nil, fmt.Errorf("environment variable %q is required", doAccessTokenEnv)
+	}
+
+	tokenSource := &tokenSource{
+		AccessToken: token,
+	}
+
+	oauthClient := oauth2.NewClient(oauth2.NoContext, tokenSource)
+	doClient, err = godo.New(oauthClient, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create godo client: %s", err)
+	}
+
+	return doClient, nil
+}
+
+// DOKSLBServiceValidator ...
+func (v *DOKSLBServiceValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	svc := &v1.Service{}
+	v.Log.V(10).Info("decoding received request")
+	err := v.decoder.Decode(req, svc)
+	if err != nil {
+		v.Log.Error(err, "bad request")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	v.Log.V(10).Info("checking received request")
+	if svc.Spec.Type != v1.ServiceTypeLoadBalancer {
+		return admission.Allowed("")
+	}
+
+	// check if request is for LB deletion
+	if svc.Spec.Type == v1.ServiceTypeLoadBalancer && svc.DeletionTimestamp != nil {
+		return admission.Allowed("the loadbalancer is deleting")
+	}
+
+	// initialize DO Client if not already initialized
+	if v.gClient == nil {
```

Comment:
I suggested this change since we do _not_ initialize the doclient in main and any call to Handle would overwrite the dummy client we initialized in our test. The only thing initialized in main, iirc, is the Log.

If we prefer to have doclient initialization in main, then this would remove the need for the check in it's entirety and we can just use v.gClient.

P.S.
> One missing bit seems to be that we do not store our initialized client in the DOKSLBServiceValidator struct.

Isn't it stored down below in the bottom of this condition via `v.gClient = doClient`?

Reply:
Yes, but it should be stored on the struct at bootstrapping time.